doc/callbacks.md

```diff
@@ -0,0 +1,484 @@
+
+| **type** | **name** | **description** |
+|----------|----------|-----------------|
+| `OptimizerType` | **`optimizer`** | The optimizer used to update the function. |
+| `FunctionType` | **`function`** | The function to be optimized. |
+| `MatType` | **`coordinates`** | The current function parameter. |
+| `size_t` | **`constraint`** | The index of the constraint. |
+| `double` | **`constraintValue`** | Constraint value of the current point. |
+
+### Gradient
+
+ Called after any call to `Gradient()`.
+
+ * `Gradient(`_`optimizer, function, coordinates, gradient`_`)`
+
+#### Attributes
+
+| **type** | **name** | **description** |
+|----------|----------|-----------------|
+| `OptimizerType` | **`optimizer`** | The optimizer used to update the function. |
+| `FunctionType` | **`function`** | The function to be optimized. |
+| `MatType` | **`coordinates`** | The current function parameter. |
+| `GradType` | **`gradient`** | Matrix that holds the gradient. |
+
+### GradientConstraint
+
+ Called after any call to `GradientConstraint()`.
+
+ * `GradientConstraint(`_`optimizer, function, coordinates, constraint, gradient`_`)`
+
+#### Attributes
+
+| **type** | **name** | **description** |
+|----------|----------|-----------------|
+| `OptimizerType` | **`optimizer`** | The optimizer used to update the function. |
+| `FunctionType` | **`function`** | The function to be optimized. |
+| `MatType` | **`coordinates`** | The current function parameter. |
+| `size_t` | **`constraint`** | The index of the constraint. |
+| `GradType` | **`gradient`** | Matrix that holds the gradient. |
+
+### BeginEpoch
+
+Called at the beginning of a pass over the data. The objective may be exact or
+an estimate depending on `exactObjective` value.
+
+ * `BeginEpoch(`_`optimizer, function, coordinates, epoch, objective`_`)`
+
+#### Attributes
+
+| **type** | **name** | **description** |
+|----------|----------|-----------------|
+| `OptimizerType` | **`optimizer`** | The optimizer used to update the function. |
+| `FunctionType` | **`function`** | The function to be optimized. |
+| `MatType` | **`coordinates`** | The current function parameter. |
+| `size_t` | **`epoch`** | The index of the current epoch. |
+| `double` | **`objective`** | Objective value of the current point. |
+
+### BeginEpoch
+
+Called at the end of a pass over the data. The objective may be exact or
+an estimate depending on `exactObjective` value.
+
+ * `EndEpoch(`_`optimizer, function, coordinates, epoch, objective`_`)`
+
+#### Attributes
+
+| **type** | **name** | **description** |
+|----------|----------|-----------------|
+| `OptimizerType` | **`optimizer`** | The optimizer used to update the function. |
+| `FunctionType` | **`function`** | The function to be optimized. |
+| `MatType` | **`coordinates`** | The current function parameter. |
+| `size_t` | **`epoch`** | The index of the current epoch. |
+| `double` | **`objective`** | Objective value of the current point. |
+
+## Built-in Callbacks
```

Comment:
I'm with Ryan on this.  From an "average use case" perspective, I think it would be better to document the built-in callbacks first, and then explain how to make customised callbacks.